x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,296 @@
+                if (newCacheFile.acquire(this)) {
+                    final CacheFile previousCacheFile = cacheFile.getAndSet(newCacheFile);
+                    assert previousCacheFile == null;
+                    return newCacheFile;
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public void onEviction(final CacheFile evictedCacheFile) {
+            try (ReleasableLock ignored = cacheEvictionLock.acquire()) {
+                if (cacheFile.compareAndSet(evictedCacheFile, null)) {
+                    evictedCacheFile.release(this);
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            if (closed.compareAndSet(false, true)) {
+                try (ReleasableLock ignored = cacheEvictionLock.acquire()) {
+                    final CacheFile currentCacheFile = cacheFile.getAndSet(null);
+                    if (currentCacheFile != null) {
+                        currentCacheFile.release(this);
+                    }
+                }
+            }
+        }
+
+        @Override
+        protected void readInternal(final byte[] buffer, final int offset, final int length) throws IOException {
+            final long position = getFilePointer() + this.offset;
+
+            int bytesRead = 0;
+            while (bytesRead < length) {
+                final long pos = position + bytesRead;
+                final int off = offset + bytesRead;
+                final int len = length - bytesRead;
+
+                CacheFile cacheFile = null;
+                try {
+                    cacheFile = getOrAcquire();
+                    if (cacheFile == null) {
+                        throw new AlreadyClosedException("Cache file evicted soon after acquisition");
+                    }
+
+                    try (ReleasableLock ignored = cacheAcquireLock.acquire()) {
+                        final FileChannel channel = cacheFile.getChannel();
+                        if (channel == null) {
+                            throw new AlreadyClosedException("Cache file evicted before accessing the file channel");
+                        }
+                        bytesRead += cacheFile.fetchRange(pos,
+                            (start, end) -> readCacheFile(channel, start, end, pos, buffer, off, len),
+                            (start, end) -> writeCacheFile(channel, start, end))
+                            .get();
+                    }
+                } catch (Exception e) {
+                    if (isEvictionException(e) == false) {
+                        throw new IOException("Fail to read data from cache", e);
+                    }
+                    // cache file was evicted during the range fetching, read bytes directly from source
+                    bytesRead += copySource(pos, pos + len, buffer, off);
+
+                } finally {
+                    if (bytesRead >= length) {
```

Comment:
I think clones need to release the last-acquired cache file if we bail out of this method by throwing an exception before completing the read?

Reply:
If the exception is an instance of AlreadyClosedException it should be released as part of the eviction mechanism. But if it's not, then it need to be released I agree.

I pushed 46684563802bc5a04a4d0756c6ccd0804a7c1c52 to always release cache files for clones in case of an exception is thrown.